Authenticate connection IDs (#3499)
quicwg/base-drafts: draft-ietf-quic-transport.md

Upon first receiving an Initial or Retry packet from the server, the client uses
the Source Connection ID supplied by the server as the Destination Connection ID
for subsequent packets, including any 0-RTT packets.  A client changes the
Destination Connection ID field if it receives a Retry packet.  A client always
changes the Destination Connection ID field in response to the first Initial
packet received from a server.  Once a client has received an Initial packet
from the server, it MUST discard any packet it receives with a different Source
Connection ID.

> Each endpoint includes the connection ID it includes in Initial packets in the

Reviewer: I think this text should be more specific about which connection ID it's talking about (the source connection ID). Also, a sentence spelling out how validation is performed can't hurt.

Author: Thanks David.  Fixed; and you get two sentences.

Upon first receiving an Initial or Retry packet from the server, the client uses
the Source Connection ID supplied by the server as the Destination Connection ID
for subsequent packets, including any 0-RTT packets.  A client changes the
Destination Connection ID field if it receives a Retry packet.  A client always
changes the Destination Connection ID field in response to the first Initial
packet received from a server.  Once a client has received an Initial packet
from the server, it MUST discard any packet it receives with a different Source
Connection ID.

> Each endpoint includes the Source Connection ID it includes in Initial packets

Reviewer 2: ```suggestion
Each endpoint includes its Initial packet Source Connection ID 
```

Author: Thanks.  I've tried to avoid that sort of train of nouns in the past.  I have an alternative for you.  More words again :(, but maybe clearer.

Reviewer 2: No worries! Just a suggestion in reaction to redundant "includes."

packets it sends in the handshake_connection_id transport parameter; see
{{transport-parameter-definitions}}.  Each endpoint validates that the value
received from the peer is identical to the value of the transport parameter.
Absense of the handshake_connection_id transport parameter or a mismatch in
values MUST be treated as a connection error of type PROTOCOL_VIOLATION.  When
sending a Retry packet or the first Initial packet, a server MUST select values
for the Source Connection ID field that differ from the values the client
includes in the Destination Connection ID field.  These measures ensure that the
choice of connection ID cannot be influenced by an attacker.

> 

Reviewer 3: I have some difficulty parsing that in the retry scenario. is "the Source Connection ID from Initial
packet" the value found in the retry packet sent by the server, or is the source connection ID in the first initial packet of the server? Currently, there is no requirement that these two be the same value.

Connection ID field it receives in the original Initial packet from the client
in the original_connection_id transport parameter. After sending a Retry
packet, a server also includes the Source Connection ID field from the Retry
packet in the retry_connection_id transport parameter.

Each endpoint ensures that the values of transport parameters matches the
values. The values provided by a peer for these transport parameters MUST match
values that an endpoint used in the Destination Connection ID field of an
Initial packet. Including connection ID values in transport parameters and
verifying them ensures that the choice of connection ID cannot be influenced by
an attacker. An endpoint MUST treat any of the following as a connection error
of type PROTOCOL_VIOLATION:

* absence of the handshake_connection_id transport parameter from either
  endpoint,

* absence of the original_connection_id transport parameter from the server,

* absence of the retry_connection_id transport parameter from the server after
  receiving a Retry packet,

* presence of the retry_connection_id transport parameter when no Retry packet
  was received, or

> * a mismatch between values of these transport parameters and the Destination

Reviewer: I think you mean `a mismatch between values of these transport parameters and the respective Connection ID fields an endpoint used.` since it doesn't just protect destination

({{migration}}); see {{issue-cid}} for details.

## Authenticating Connection IDs {#cid-auth}

The choice each endpoint makes about connection IDs during the handshake is
authenticated by including all values in transport parameters; see
{{transport-parameters}}. This ensures that all connection IDs are authenticated
by the cryptographic handshake.

Each endpoint includes the value of the Source Connection ID field from Initial

> packets it sends in the handshake_connection_id transport parameter; see

Reviewer 4: One could read that this Connection ID is only used on Initial packets.  If so, I think initial_connection_id would be a better TP name.  But I think this CID can be used used on Initial and 0-RTT from the client and all Initial, Handshake and 1-RTT from the server?

Reviewer: I think it's called `handshake_connection_id` because it is the connection ID used during the handshake, I think it's a reasonable name

Reviewer 5: In keeping with retry_connection_id being the SCID that is used on Retry packets, I have a preference for initial_connection_id, since it is the SCID used on Initial packets. 

In the same vein, and this is a weaker preference, I would propose changing original_connection_id to original_initial_connection_id. I think this makes it clear that this might have been initial_connection_id had it not been for a Retry.

{{transport-parameter-definitions}}. A server includes the Destination
Connection ID field it receives in the original Initial packet from the client
in the original_connection_id transport parameter. After sending a Retry
packet, a server also includes the Source Connection ID field from the Retry
packet in the retry_connection_id transport parameter.

Each endpoint ensures that the values of transport parameters matches the

> values. The values provided by a peer for these transport parameters MUST match

Reviewer 6: This sentence feels like it ends prematurely.  Perhaps...
```suggestion
Each endpoint ensures that the values of the peer's transport parameters
match the values the endpoint sent in its packet headers. The values
provided by a peer for these transport parameters MUST match the
```

values that an endpoint used in the Destination Connection ID field of an
Initial packet. Including connection ID values in transport parameters and
verifying them ensures that the choice of connection ID cannot be influenced by
an attacker. An endpoint MUST treat any of the following as a connection error
of type PROTOCOL_VIOLATION:

* absence of the handshake_connection_id transport parameter from either
  endpoint,

* absence of the original_connection_id transport parameter from the server,

* absence of the retry_connection_id transport parameter from the server after
  receiving a Retry packet,

* presence of the retry_connection_id transport parameter when no Retry packet
  was received, or

* a mismatch between values received from a peer in these transport parameters
  and the value sent in the corresponding Destination Connection ID fields of
  Initial packets.

> If a zero-length connection IDs is selected, the corresponding transport

Reviewer 4: ```suggestion
If a zero-length connection ID is selected, the corresponding transport
```

({{migration}}); see {{issue-cid}} for details.

## Authenticating Connection IDs {#cid-auth}

The choice each endpoint makes about connection IDs during the handshake is
authenticated by including all values in transport parameters; see
{{transport-parameters}}. This ensures that all connection IDs used for the
handshake are also authenticated by the cryptographic handshake.

> Each endpoint includes the value of the Source Connection ID field from the most

Reviewer: Should this be the "first packet it sent" as opposed to "most recent" for consistency?

> Each endpoint includes the value of the Source Connection ID field from Initial
> packets it sends in the handshake_connection_id transport parameter; see

Reviewer 5: ```suggestion
Each endpoint includes the value of the Source Connection ID field from the 
most recent Initial packet it sent in the handshake_connection_id transport
parameter; see
```

parameter; see {{transport-parameter-definitions}}. A server includes the

> Destination Connection ID field it receives in original Initial packets from the

Reviewer: I would rephrase this to "A server includes the Destination Connection ID field from the first Initial packet it received from the client in the original_connection_id transport parameter; note that if the server sent a Retry packet this refers to the first Initial packet received before sending the Retry packet."

Reviewer 3: We already have a definition for an original_connection_id transport parameter (see section  17.2.5.3. of the transport spec, Continuing a Handshake After Retry). From the point of view of the client, we have the following:

1) Without retry:
* Initial DCID: DCID of the first Initial packet sent by the client.
* Server SCID: CID chosen by the server to identify the connection during the handshake. Also used by the client for further Initial or Handshake packets once the client receives a first handshake message from the server.

2) With Retry:
* Original DCID: the DCID in the first Initial packet sent by the client, when that packet triggered a Retry.
* Server suggested DCID: the SCID of the Retry packet sent by the server. 
* Initial DCID: the DCID in the first Initial packet sent by the client in response to the Retry. MUST be set to the Server suggested DCID.
* Server SCID: the SCID chosen by the server to identify the connection. May or may not be the same as Server suggested DCID.

Maybe we should tighten these definitions.

Reviewer: `original_connection_id` used to only apply when a Retry was involved. This PR changes that: servers now always send `original_connection_id` no matter what.

client - Initial packets received by the server prior to sending a Retry packet

> - in the original_connection_id transport parameter. After sending a Retry

Reviewer: I would s/After sending a/If it sent a/

packet, a server also includes the Source Connection ID field from the Retry
packet in the retry_connection_id transport parameter.

The values provided by a peer for these transport parameters MUST match the

> values that an endpoint used in the Destination Connection ID field of Initial

Reviewer: This sentence appears to apply to all 3 transport parameters, but it mentions Destination CID. I would s/Destination/Source and Destination/

packets that it sent. Including connection ID values in transport parameters and
verifying them ensures that that an attacker cannot influence the choice of
connection ID for a successful connection by injecting packets carrying
attacker-chosen connection IDs during the handshake. An endpoint MUST
treat any of the following as a connection error of type PROTOCOL_VIOLATION:

* absence of the initial_connection_id transport parameter from either
  endpoint,

* absence of the original_connection_id transport parameter from the server,

* absence of the retry_connection_id transport parameter from the server after
  receiving a Retry packet,

* presence of the retry_connection_id transport parameter when no Retry packet
  was received, or

* a mismatch between values received from a peer in these transport parameters
  and the value sent in the corresponding Destination Connection ID fields of

>   Initial packets.

Reviewer 6: ```suggestion
  Initial packets or the Source Connection ID field of a Retry packet.
```
...though perhaps this is fine, since the client will then use this value as its DCID for the next Initial packet.

Author: > the client will then use this value as its DCID for the next Initial

Yeah, that was my thought.  The Retry.SCID thing is a choice people are free to use, but that doesn't need to be written down.  This way is simpler.

handshake are also authenticated by the cryptographic handshake.

Each endpoint includes the value of the Source Connection ID field from the
first Initial packet it sent in the initial_source_connection_id transport
parameter; see {{transport-parameter-definitions}}. A server includes the
Destination Connection ID field from the first Initial packet it received from
the client in the original_destination_connection_id transport parameter; if
the server sent a Retry packet this refers to the first Initial packet received
before sending the Retry packet. If it sends a Retry packet, a server also
includes the Source Connection ID field from the Retry packet in the
retry_source_connection_id transport parameter.

The values provided by a peer for these transport parameters MUST match the
values that an endpoint used in the Destination and Source Connection ID fields
of Initial packets that it sent. Including connection ID values in transport
parameters and verifying them ensures that that an attacker cannot influence
the choice of connection ID for a successful connection by injecting packets
carrying attacker-chosen connection IDs during the handshake. An endpoint MUST

> treat any of the following as a connection error of type PROTOCOL_VIOLATION:

Reviewer: I'm not sure I agree. TRANSPORT_PARAMETER_ERROR indicates a parse issue, whereas this validation is performed somewhere else in our implementation.

> treat any of the following as a connection error of type PROTOCOL_VIOLATION:

Reviewer 7: Our rule for FRAME_ENCODING_ERROR is that we use that error code when the error is detectable while parsing the frame, without accessing connection state. I'd argue that we should apply a similar logic here. The absence of `initial_source_connection_id` and `original_destination_connection_id` definitely falls in that category, and I'd put the absence / presence of `retry_source_connection_id` in the same category, although I could see why someone could make an argument against this.

The mismatch seems to be a clear case for PROTOCOL_VIOLATION.

> treat any of the following as a connection error of type PROTOCOL_VIOLATION:

Author: I like that rationale.  I could also see my way to choose TRANSPORT_PARAMETER_ERROR.  We use that for transport parameters that are present when they are disallowed (like preferred_address from a client) in addition to obvious encoding problems, which suggests that you might implement this as `if transport_parameters.bad() then TRANSPORT_PARAMETER_ERROR`, but this validation does involve accessing external state, as you say, so a different code is fully justified.

  packet was received, or

* a mismatch between values received from a peer in these transport parameters
  and the value sent in the corresponding Destination Connection ID fields of
  Initial packets.

If a zero-length connection ID is selected, the corresponding transport
parameter is included with a zero-length value.

{{fig-auth-cid}} shows the connection IDs that are used in a complete
handshake. The exchange of Initial and Retry packets is shown, plus the later
exchange of 1-RTT packets that includes the connection ID established during
the handshake.

~~~
Client                                                  Server

Initial: DCID=S1, SCID=C1 ->
                                   <- Retry*: DCID=C1, SCID=S2
Initial*: DCID=S2, SCID=C1 ->
                                  <- Initial: DCID=C1, SCID=S3
                             ...
1-RTT: DCID=S3 ->
                                             <- 1-RTT: DCID=C1
~~~
{: #fig-auth-cid title="Use of Connection IDs in a Handshake"}

For the handshake in {{fig-auth-cid}} the client sets the value of the
initial_source_connection_id transport parameter to `C1`. If the server sends a

> Retry packet (that is, the packets marked with a '*' are sent), it sets

Reviewer: I really like the diagram, but I would prefer to split it in two. There are two possible handshakes described here: without and with retry involved. I think that the `the packets marked with a '*' are sent` caveat is confusing, because when those packets aren't sent, the `1-RTT: DCID=S3` needs to be replaced by `1-RTT: DCID=S2`. It might be simpler to have two diagrams and two examples.

parameters extension.  An endpoint SHOULD treat receipt of duplicate transport
parameters as a connection error of type TRANSPORT_PARAMETER_ERROR.

> A server MUST include the original_destination_connection_id transport parameter

Reviewer 7: This is inconsistent with the error condition above:

> * absence of the original_destination_connection_id transport parameter from the server"

Author: Not following.  These are consistent by my reading:

> A server MUST include the original_destination_connection_id transport parameter

> An endpoint MUST
treat any of the following as a connection error of type PROTOCOL_VIOLATION:
> * absence of the original_destination_connection_id transport parameter from the server

Reviewer: @martinthomson you truncated the sentence in your quote. The full sentence from your PR is:

> A server MUST include the original_destination_connection_id transport parameter
> ({{transport-parameter-definitions}}) if it sent a Retry packet to enable
> validation of the Retry, as described in {{packet-retry}}.

How about we keep the first line, but replace the rest:

> A server MUST include the original_destination_connection_id transport parameter
> to mitigate an attacker's ability to tamper with connection IDs during the handshake.

Author: Ahh, that's lazy of me.  Yes, that is good.  Thanks.

Author: Oh, that's a throwback paragraph.  I'm going to replace that with a more generic reference.

not store a transport parameter it cannot process.

A client MUST NOT use remembered values for the following parameters:

> ack_delay_exponent, handshake_connection_id, original_connection_id,

Reviewer: Shouldn't `original_connection_id` stay in this list? And what about `handshake_connection_id`?

* initial_max_stream_data_uni
* initial_max_streams_bidi
* initial_max_streams_uni

> * active_connection_id_limit

Reviewer 6: Why was this removed in this PR?

Author: A rebasing error.  I added something to this list, sorted it, removed it, now fixed it.

processing a Retry packet; {{packet-0rtt}} contains more information on this.

A server acknowledges the use of a Retry packet for a connection using the
retry_connection_id transport parameter; see

> {{transport-parameter-definitions}}. If the server sends a Retry packet, it

Reviewer: We may want to rephrase this to more clearly state that `original_connection_id` is always filled in by the server

processing a Retry packet; {{packet-0rtt}} contains more information on this.

A server acknowledges the use of a Retry packet for a connection using the
retry_connection_id transport parameter; see
{{transport-parameter-definitions}}.  If it sends a Retry packet, the server
also subsequently includes the value of the Source Connection ID field from
the Retry packet in its retry_connection_id transport parameter.

> A server always includes an original_connection_id transport parameter. If it

Reviewer: I think we can remove this entire paragraph as it appears to be redundant with other parts of this PR, and it seems to have discrepancies.

processing a Retry packet; {{packet-0rtt}} contains more information on this.

A server acknowledges the use of a Retry packet for a connection using the
retry_source_connection_id transport parameter; see
{{transport-parameter-definitions}}.  If the server sends a Retry packet, it
also subsequently includes the value of the Source Connection ID field from
the Retry packet in its retry_source_connection_id transport parameter.

If the client received and processed a Retry packet, it MUST validate that the
retry_source_connection_id transport parameter is present and correct;
otherwise, it MUST validate that the transport parameter is absent. A client
MUST treat a failed validation as a connection error of type

> TRANSPORT_PARAMETER_ERROR.

Reviewer: This contradicts line 1568
> An endpoint MUST treat any of the following as a connection error of type PROTOCOL_VIOLATION

I have a slight preference for PROTOCOL_VIOLATION over TRANSPORT_PARAMETER_ERROR, though I don't much care as long as this section matches the other one.

: The value of the Destination Connection ID field from the first Initial packet
  sent by the client.  This transport parameter is only sent by a server.  This

>   is the same value sent in the "Original Destination Connection ID" field that

Reviewer: Thanks for catching that, my bad for missing it.

original_connection_id (0x00):

: The value of the Destination Connection ID field from the first Initial packet
  sent by the client; see {{cid-auth}}.  This transport parameter is only sent

>   by a server.  The value of this parameter depends on whether the server

Reviewer: I don't understand the sentence "The value of this parameter depends on whether the server sent a Retry packet". From reading the rest of this PR it sounds like this TP has the same value no matter what happened with RETRY. Am I missing something?

  When a zero-length connection ID is being used, the active_connection_id_limit
  parameter MUST NOT be sent.

handshake_connection_id (0x0f):

: The value that the endpoint included in the Source Connection ID field of the
  first Initial packet it sends during the handshake. Endpoints MUST validate
  that this transport parameter is present and that it matches the value that
  was received in Initial packets. Authenticating this value ensures that an

>   attacker is unable to influence the selection of connection IDs during the

Reviewer: to avoid jinxing things, how about `Authenticating this value mitigates an attacker's ability to influence the selection of connection IDs during the handshake.`

Reviewer 8: What if some of these CIDs are zero length? Do we omit them?

Reviewer: My read of the zero-length CID case is that the TP MUST be included, and be zero-length. Perhaps we should add a sentence to clarify that.

Author: Done! Thanks both.

As for the "mitigates" formulation, I think that we are safe in making claims, but I realize that we don't need to.  I've removed a lot of text from the transport parameter definitions and replaced it with a pointer to the new section.

  When a zero-length connection ID is being used, the active_connection_id_limit
  parameter MUST NOT be sent.

initial_connection_id (0x0f):

: The value that the endpoint included in the Source Connection ID field of the

>   first Initial packet it sends during the handshake; see {{cid-auth}}.

Reviewer: nit: I would s/packet it sends during the handshake/packet it sent on this connection/

Reviewer 3: Everywhere else, "initial_connection_id" is defined as _The value that the endpoint included in the  **Destination** Connection ID field of the first Initial packet it sends during the handshake._
I think that the simplest solution is to s/Source/Destination/

Reviewer: I don't think that's right @huitema. `initial_connection_id` refers to the Source Connection ID because that's the one that needs protection. The Destination Connection ID is covered by `original_connection_id `.

Reviewer 3: @DavidSchinazi If you want to refer to the source CID set by the client in its first Initial packet, you should really use a different name than `initial_connection_id`, because this is widely used to refer to the initial destination CID. Maybe you should call that `initial_source_connection_id`. But using the same name for 2 concepts seems bad.

Reviewer: I didn't pick the name, but I agree that `initial_source_connection_id` might be clearer than `initial_connection_id`.
